azure/scope/machinepoolmachine.go

```diff
@@ -261,6 +274,152 @@ func (s *MachinePoolMachineScope) UpdateStatus(ctx context.Context) error {
+	if s.isNodeDrainAllowed() {
+		patchHelper, err := patch.NewHelper(s.AzureMachinePoolMachine, s.client)
+		if err != nil {
+			return errors.Wrap(err, "failed to build a patchHelper when draining node")
+		}
+
+		s.V(4).Info("Draining node", "node", s.AzureMachinePoolMachine.Status.NodeRef.Name)
+		// The DrainingSucceededCondition never exists before the node is drained for the first time,
+		// so its transition time can be used to record the first time draining.
+		// This `if` condition prevents the transition time to be changed more than once.
+		if conditions.Get(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition) == nil {
+			conditions.MarkFalse(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition, clusterv1.DrainingReason, clusterv1.ConditionSeverityInfo, "Draining the node before deletion")
+		}
+
+		if err := patchHelper.Patch(ctx, s.AzureMachinePoolMachine); err != nil {
+			return errors.Wrap(err, "failed to patch AzureMachinePoolMachine")
+		}
+
+		if err := s.drainNode(ctx, node); err != nil {
+			conditions.MarkFalse(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition, clusterv1.DrainingFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+			return err
+		}
+
+		conditions.MarkTrue(s.AzureMachinePoolMachine, clusterv1.DrainingSucceededCondition)
+	}
+
+	return nil
+}
+
+func (s *MachinePoolMachineScope) drainNode(ctx context.Context, node *corev1.Node) error {
+	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolMachineScope.drainNode")
+	defer span.End()
+
+	restConfig, err := remote.RESTConfig(ctx, MachinePoolMachineScopeName, s.client, client.ObjectKey{
+		Name:      s.ClusterName(),
+		Namespace: s.AzureMachinePoolMachine.Namespace,
+	})
+
+	if err != nil {
+		s.Error(err, "Error creating a remote client while deleting Machine, won't retry")
+		return nil
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		s.Error(err, "Error creating a remote client while deleting Machine, won't retry")
+		return nil
+	}
+
+	drainer := &drain.Helper{
+		Client:              kubeClient,
+		Force:               true,
+		IgnoreAllDaemonSets: true,
+		DeleteLocalData:     true,
+		GracePeriodSeconds:  -1,
+		// If a pod is not evicted in 20 seconds, retry the eviction next time the
+		// machine gets reconciled again (to allow other machines to be reconciled).
+		Timeout: 20 * time.Second,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			s.V(4).Info(fmt.Sprintf("%s pod from Node", verbStr),
+				"pod", fmt.Sprintf("%s/%s", pod.Name, pod.Namespace))
+		},
+		Out:    writer{klog.Info},
+		ErrOut: writer{klog.Error},
+		DryRun: false,
+	}
+
+	if noderefutil.IsNodeUnreachable(node) {
+		// When the node is unreachable and some pods are not evicted for as long as this timeout, we ignore them.
+		drainer.SkipWaitForDeleteTimeoutSeconds = 60 * 5 // 5 minutes
```

Comment:
nit
```suggestion
		drainer.SkipWaitForDeleteTimeoutSeconds = int(5 * time.Minute)
```

Reply:
I don't think that will do what you think it does: https://play.golang.org/p/yk9BWxHPCk-

Reply:
Ah, You're right. Thanks for the explanation!